### ISSUE TITLE
Can NVIDIA support running more than one model inference from one GPU

### ISSUE THREAD

**Issue by aisensiy** [NVIDIA/dl-inference-server#7]
It seems that right now one GPU can only be used by one docker. What if I want to serve more than one model at one GPU?

**Comment by deadeyegoodwin** [comment]
By "used by one docker" do you mean that the inference server container can only see one GPU. In general that is not true though in certain orchestration environments (like kubernetes) that might be true. The inference server can manage all the GPUs on a system. So if your workstation has multiple GPUs and you run the inference server container there it will see all the GPUs and you can have it perform inference on those GPUs.

I think you are also asking about having multiple models on a GPU. Yes, the inference server by default will use the GPU for all models from the model store. The default is to have one model instance on each GPU. So if you have 4 models in your model store and 2 GPUs in your system then the inference server will create 2 instances of each model, one for each GPU. That means that it will be able to simultaneously handle 8 inference requests (assuming there are 2 requests for each model). You can change this default and control exactly whick models run on which GPUs by changing the model configuration in the model store. Look in model_config.proto but the description of InstanceGroup. We are improving the documentation around this feature in future docs. 

Please follow-up here if you have additional questions.

**Comment by aisensiy** [comment]
Thanks for your reply.

So you mean I can hold more than one model in one GPU if I deployment multiple models in one docker?

**Comment by deadeyegoodwin** [comment]
Yes, you can have multiple models being served by the same inference server. When you set up your model store simply include multiple models within the model store and the inference server will make them all available. Currently, the only limit on the number of models is GPU memory (since each model will require some amount of GPU memory). In the future we plan to remove this limitation.

**Comment by GoodJoey** [comment]
@deadeyegoodwin 
yes, we can have multiple modes inference in one GPU(if the gpu memory is enough), but seems the actual usage of the GPU will be affected.
For example, i have mode A and B inference, both of them will take 10 ms for each request if they run on separate GPU (qps is 100), and the GPU usage are all about 30%. 
Then, I tried to run both A and B in one GPU to increase the GPU usage, the ideal result is the time spent should be still around 10 ms for both A and B (qps becomes 100 *2 = 200), but the actual result is, the time raises to 30 ms for both A and B (qps is 33 * 2 = 66).
do you know why is that? is there any way to solve this problem?

**Comment by deadeyegoodwin** [comment]
What are some characteristics of your models? What framework? Can you share them? How are you measuring these performance numbers? If each model was fulling occupying the GPU (or I/O) then you could see the QPS for two models not be any better than for one (that is, you could see total QPS for 2 models of 100), but it is unexpected that the QPS would drop so much relative to one model.

**Comment by GoodJoey** [comment]
If each model was fulling occupying the GPU (or I/O), according to nvidia-smi the GPU usage is 30%, the gpu mem is also enough to put A and B in one gpu. And the cpu/IO also was not the bottleneck.

I did some search, seems GPU use round-robin to schedule the process, A and B is not actual run concurrently and plus some context-switching cost, i think that's the reason the qps drop from 100 to 66.

BTW, my GPU is 1080ti.

**Comment by deadeyegoodwin** [comment]
GPU will time-slice if multiple processes are attempting to use the GPU at the same time. The inference server is a single process, so assuming you don't have another process running and using the GPU, the inference server will not incur and GPU context-switch overhead. Inference server uses CUDA streams to simultaneously execute multiple models. You can read some about it here: https://docs.nvidia.com/deeplearning/sdk/inference-user-guide/index.html#concurrent-model

I don't have an explanation for your results which is why I asked about the model, framework, etc.

**Comment by GoodJoey** [comment]
ok， so you mean the qps should stay around 100（A 50， B 50） if there is no other bottleneck.
i run the inference of A and B in different docker container, so it should be two different processes?
And my framework is mxnet.
Thanks for your information, i will do some experience in tensorRT.

**Comment by deadeyegoodwin** [comment]
It is difficult to say what you should expect for performance without knowing more about your setup. Having 2 containers sharing the GPU is not ideal. Have you tried running one inference server container and have that one container serve both models? That will give you the best GPU utilization.
